File: sdks/php/src/Module/TransactionManager.php
<?php

declare(strict_types=1);

namespace Synap\SDK\Module;

use Synap\SDK\SynapClient;

/**
 * Transaction operations (Redis-compatible)
 */
class TransactionManager
{
    public function __construct(
        private SynapClient $client
    ) {
    }

    /**
     * Start a transaction (MULTI)
     *
     * @param string|null $clientId Optional client identifier to group commands within the same transaction
     * @return array{success: bool, message: string}
     */
    public function multi(?string $clientId = null): array
    {
        $payload = [];
        if ($clientId !== null) {
            $payload['client_id'] = $clientId;
        }

        $response = $this->client->execute('transaction.multi', '', $payload);
        $payload = $response['payload'] ?? $response;

        return [
            'success' => $payload['success'] ?? true,
            'message' => $payload['message'] ?? 'Transaction started',
        ];
    }

    /**
     * Discard the current transaction (DISCARD)
     *
     * @param string|null $clientId Optional client identifier for the transaction
     * @return array{success: bool, message: string}
     */
    public function discard(?string $clientId = null): array
    {
        $payload = [];
        if ($clientId !== null) {
            $payload['client_id'] = $clientId;
        }

        $response = $this->client->execute('transaction.discard', '', $payload);
        $payload = $response['payload'] ?? $response;

        return [
            'success' => $payload['success'] ?? true,
            'message' => $payload['message'] ?? 'Transaction discarded',
        ];
    }

    /**
     * Watch keys for optimistic locking (WATCH)
     *
     * @param list<string> $keys List of keys to watch for changes
     * @param string|null $clientId Optional client identifier for the transaction
     * @return array{success: bool, message: string}
     * @throws \InvalidArgumentException If keys list is empty
     */
    public function watch(array $keys, ?string $clientId = null): array
    {
        if (empty($keys)) {
            throw new \InvalidArgumentException('Transaction watch requires at least one key');
        }

        $payload = ['keys' => $keys];
        if ($clientId !== null) {
            $payload['client_id'] = $clientId;
        }

        $response = $this->client->execute('transaction.watch', '', $payload);
        $payload = $response['payload'] ?? $response;

        return [
            'success' => $payload['success'] ?? true,
            'message' => $payload['message'] ?? 'Keys watched',
        ];
    }

    /**
     * Remove all watched keys (UNWATCH)
     *
     * @param string|null $clientId Optional client identifier for the transaction
     * @return array{success: bool, message: string}
     */
    public function unwatch(?string $clientId = null): array
    {
        $payload = [];
        if ($clientId !== null) {
            $payload['client_id'] = $clientId;
        }

        $response = $this->client->execute('transaction.unwatch', '', $payload);
        $payload = $response['payload'] ?? $response;

        return [
            'success' => $payload['success'] ?? true,
            'message' => $payload['message'] ?? 'Keys unwatched',
        ];
    }

    /**
     * Execute queued commands (EXEC)
     *
     * @param string|null $clientId Optional client identifier for the transaction
     * @return array{success: true, results: list<mixed>}|array{success: false, aborted: true, message?: string}
     */
    public function exec(?string $clientId = null): array
    {
        $payload = [];
        if ($clientId !== null) {
            $payload['client_id'] = $clientId;
        }

        $response = $this->client->execute('transaction.exec', '', $payload);
        $payload = $response['payload'] ?? $response;

        if (isset($payload['results']) && is_array($payload['results'])) {
            return [
                'success' => true,
                'results' => $payload['results'],
            ];
        }

        return [
            'success' => false,
            'aborted' => true,
            'message' => $payload['message'] ?? null,
        ];
    }
}

File: sdks/csharp/src/Synap.SDK/Modules/TransactionManager.cs
using System.Collections.ObjectModel;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Synap.SDK.Modules;

/// <summary>
/// Response from transaction operations.
/// </summary>
public class TransactionResponse
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("message")]
    public string Message { get; set; } = string.Empty;
}

/// <summary>
/// Successful transaction execution result.
/// </summary>
public class TransactionExecSuccess
{
    [JsonPropertyName("success")]
    public bool Success { get; set; } = true;

    [JsonPropertyName("results")]
    public Collection<JsonElement> Results { get; } = new();
}

/// <summary>
/// Aborted transaction execution result.
/// </summary>
public class TransactionExecAborted
{
    [JsonPropertyName("success")]
    public bool Success { get; set; } = false;

    [JsonPropertyName("aborted")]
    public bool Aborted { get; set; } = true;

    [JsonPropertyName("message")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Message { get; set; }
}

/// <summary>
/// Transaction operations (Redis-compatible).
/// </summary>
public sealed class TransactionManager
{
    private readonly SynapClient _client;

    /// <summary>
    /// Initializes a new instance of the <see cref="TransactionManager"/> class.
    /// </summary>
    /// <param name="client">The Synap client instance.</param>
    public TransactionManager(SynapClient client)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
    }

    /// <summary>
    /// Start a transaction (MULTI).
    /// </summary>
    /// <param name="clientId">Optional client identifier to group commands within the same transaction.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Transaction response with success status and message.</returns>
    public async Task<TransactionResponse> MultiAsync(
        string? clientId = null,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>();
        if (!string.IsNullOrWhiteSpace(clientId))
        {
            data["client_id"] = clientId;
        }

        using var response = await _client.ExecuteAsync("transaction.multi", string.Empty, data, cancellationToken).ConfigureAwait(false);
        if (response.RootElement.TryGetProperty("payload", out var payload))
        {
            return JsonSerializer.Deserialize<TransactionResponse>(payload.GetRawText()) ?? new TransactionResponse
            {
                Success = true,
                Message = "Transaction started"
            };
        }

        return new TransactionResponse
        {
            Success = true,
            Message = "Transaction started"
        };
    }

    /// <summary>
    /// Discard the current transaction (DISCARD).
    /// </summary>
    /// <param name="clientId">Optional client identifier for the transaction.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Transaction response with success status and message.</returns>
    public async Task<TransactionResponse> DiscardAsync(
        string? clientId = null,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>();
        if (!string.IsNullOrWhiteSpace(clientId))
        {
            data["client_id"] = clientId;
        }

        using var response = await _client.ExecuteAsync("transaction.discard", string.Empty, data, cancellationToken).ConfigureAwait(false);
        if (response.RootElement.TryGetProperty("payload", out var payload))
        {
            return JsonSerializer.Deserialize<TransactionResponse>(payload.GetRawText()) ?? new TransactionResponse
            {
                Success = true,
                Message = "Transaction discarded"
            };
        }

        return new TransactionResponse
        {
            Success = true,
            Message = "Transaction discarded"
        };
    }

    /// <summary>
    /// Watch keys for optimistic locking (WATCH).
    /// </summary>
    /// <param name="keys">Collection of keys to watch for changes.</param>
    /// <param name="clientId">Optional client identifier for the transaction.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Transaction response with success status and message.</returns>
    /// <exception cref="ArgumentException">Thrown when keys collection is empty.</exception>
    public async Task<TransactionResponse> WatchAsync(
        IReadOnlyCollection<string> keys,
        string? clientId = null,
        CancellationToken cancellationToken = default)
    {
        if (keys == null || keys.Count == 0)
        {
            throw new ArgumentException("Transaction watch requires at least one key", nameof(keys));
        }

        var data = new Dictionary<string, object?>
        {
            ["keys"] = keys
        };
        if (!string.IsNullOrWhiteSpace(clientId))
        {
            data["client_id"] = clientId;
        }

        using var response = await _client.ExecuteAsync("transaction.watch", string.Empty, data, cancellationToken).ConfigureAwait(false);
        if (response.RootElement.TryGetProperty("payload", out var payload))
        {
            return JsonSerializer.Deserialize<TransactionResponse>(payload.GetRawText()) ?? new TransactionResponse
            {
                Success = true,
                Message = "Keys watched"
            };
        }

        return new TransactionResponse
        {
            Success = true,
            Message = "Keys watched"
        };
    }

    /// <summary>
    /// Remove all watched keys (UNWATCH).
    /// </summary>
    /// <param name="clientId">Optional client identifier for the transaction.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Transaction response with success status and message.</returns>
    public async Task<TransactionResponse> UnwatchAsync(
        string? clientId = null,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>();
        if (!string.IsNullOrWhiteSpace(clientId))
        {
            data["client_id"] = clientId;
        }

        using var response = await _client.ExecuteAsync("transaction.unwatch", string.Empty, data, cancellationToken).ConfigureAwait(false);
        if (response.RootElement.TryGetProperty("payload", out var payload))
        {
            return JsonSerializer.Deserialize<TransactionResponse>(payload.GetRawText()) ?? new TransactionResponse
            {
                Success = true,
                Message = "Keys unwatched"
            };
        }

        return new TransactionResponse
        {
            Success = true,
            Message = "Keys unwatched"
        };
    }

    /// <summary>
    /// Execute queued commands (EXEC).
    /// </summary>
    /// <param name="clientId">Optional client identifier for the transaction.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Transaction execution result. If successful, returns results array. If aborted (due to watched keys changed), returns aborted status.</returns>
    public async Task<object> ExecAsync(
        string? clientId = null,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>();
        if (!string.IsNullOrWhiteSpace(clientId))
        {
            data["client_id"] = clientId;
        }

        using var response = await _client.ExecuteAsync("transaction.exec", string.Empty, data, cancellationToken).ConfigureAwait(false);
        if (response.RootElement.TryGetProperty("payload", out var payload))
        {
            if (payload.TryGetProperty("results", out var results))
            {
                var successResult = new TransactionExecSuccess();
                if (results.ValueKind == JsonValueKind.Array)
                {
                    foreach (var item in results.EnumerateArray())
                    {
                        successResult.Results.Add(item);
                    }
                }
                return successResult;
            }

            if (payload.TryGetProperty("aborted", out _))
            {
                return JsonSerializer.Deserialize<TransactionExecAborted>(payload.GetRawText()) ?? new TransactionExecAborted();
            }
        }

        return new TransactionExecAborted();
    }
}

